--- a/xs/Unpacker.xs	2026-07-04 11:46:03.611353506 +0800
+++ b/xs/Unpacker.xs	2026-07-04 11:46:03.622484327 +0800
@@ -39,12 +39,8 @@
 	CODE:
 		b = SvPV (buffer, size);
 
-		available = msgpack_unpacker_buffer_capacity (&self->unpacker);
-		if (size > available)
-		{
-			STRLEN extra = size-available;
-			msgpack_unpacker_reserve_buffer (&self->unpacker, extra);
-		}
+		if (!msgpack_unpacker_reserve_buffer (&self->unpacker, size))
+			croak ("unpack: out of memory");
 
 		memcpy (msgpack_unpacker_buffer (&self->unpacker), b, size);
 		msgpack_unpacker_buffer_consumed (&self->unpacker, size);
